File: QuickSort/CPP/QuickSort.cpp
//
// Created by Shahriar Nasim Nafi on 2/7/22.
//

#include <iostream>

using namespace std;

int partition(int array[], int start, int end) {

    int p = start - 1;
    int pivot = array[end - 1];

    for (int i = start; i < end - 1; i++) {
        if (pivot > array[i]) {
            p += 1;
            int temp = array[p];
            array[p] = array[i];
            array[i] = temp;
        }
    }

    p += 1;
    int temp = array[p];
    array[p] = pivot;
    array[end - 1] = temp;
    return p;

}

void quickSort(int array[], int start, int end) {

    if (start >= end) {
        return;
    }

    int p = partition(array, start, end);

    quickSort(array, start, p);
    quickSort(array, p + 1, end);
}

File: QuickSort/Swift/QuickSort.swift
//
// Created by Shahriar Nasim Nafi on 2/7/22.
//

func partitionArray<T: Comparable>(_ array: inout [T], start: Int, end: Int) -> Int {
    let pivot = array[end - 1]
    var p = start - 1
    
    for i in start..<end - 1 {
        if array[i] < pivot {
            p+=1
            (array[p], array[i]) =  (array[i], array[p])
        }
    }
    
    p+=1
    (array[p], array[end - 1]) =  (pivot, array[p])
    
    return p
}

extension Array where Element: Comparable {
    mutating func quickSort(start: Int? = nil, end: Int? = nil) {
        let start = start ?? 0
        let end = end ?? self.count
        
        if start >= end {
            return
        }
        
        let p = partitionArray(&self, start: start, end: end)
        
        quickSort(start: start, end: p)
        quickSort(start: p + 1, end: end)
    }
}
